src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs

```diff
@@ -195,5 +195,233 @@ public static void InvokeFunctionAcceptingArrayTypes(Function objectPrototype, s
+        {
+            var factory = new Function(@"
+                return new Promise((resolve, reject) => {
+                  setTimeout(() => {
+                    resolve('foo');
+                  }, 300);
+                });");
+
+            var promise = (Task<object>)factory.Call();
+            var value = await promise;
+
+            Assert.Equal("foo", (string)value);
+            
+        }
+
+        [Fact]
+        public static async Task ResolveJSObjectPromise()
+        {
+            for (int i = 0; i < 10; i++)
+            {
+                var factory = new Function(@"
+                return new Promise((resolve, reject) => {
+                  setTimeout(() => {
+                    resolve({foo:'bar'});
+                  }, 300);
+                });");
+
+                var promise = (Task<object>)factory.Call();
+                var value = (JSObject)await promise;
+
+                Assert.Equal("bar", value.GetObjectProperty("foo"));
+
+                Runtime.InvokeJS("gc();");// needs v8 flag --expose-gc
+            }
+        }
+
+        [Fact]
+        public static async Task RejectPromise()
+        {
+            var factory = new Function(@"
+                return new Promise((resolve, reject) => {
+                  setTimeout(() => {
+                    reject('fail');
+                  }, 300);
+                });");
+
+            var promise = (Task<object>)factory.Call();
+
+            var ex = await Assert.ThrowsAsync<JSException>(async () => await promise);
+            Assert.Equal("fail", ex.Message);
+        }
+
+        [Fact]
+        public static async Task RejectPromiseError()
+        {
+            var factory = new Function(@"
+                return new Promise((resolve, reject) => {
+                  setTimeout(() => {
+                    reject(new Error('fail'));
+                  }, 300);
+                });");
+
+            var promise = (Task<object>)factory.Call();
+
+            var ex = await Assert.ThrowsAsync<JSException>(async () => await promise);
+            Assert.Equal("Error: fail", ex.Message);
+        }
+
+
+        [ActiveIssue("not implemented")]
+        [Fact]
+        public static void RoundtripPromise()
+        {
+            var factory = new Function(@"
+                var dummy=new Promise((resolve, reject) => {});
+                return {
+                    dummy:dummy,
+                    check:(promise)=>{
+                        console.log(JSON.stringify(promise));
+                        return promise===dummy ? 1:0;
+                    }
+                }");
+
+            var obj = (JSObject)factory.Call();
+            var dummy = obj.GetObjectProperty("dummy");
+            Assert.IsType<Task<object>>(dummy);
+            var check = obj.Invoke("check", dummy);
+            Assert.Equal(1, check);
+        }
+
+
+        [Fact]
+        public static async Task ResolveTask()
```

Comment:
It's hard for me to tell what's going on in this test or what it's verifying. Please use more descriptive names and/or add comments, for example, it's not clear to me why check should be 0 in the second assert